File: 3.longest-substring-without-repeating-characters.ts
/*
 * @lc app=leetcode id=3 lang=typescript
 *
 * [3] Longest Substring Without Repeating Characters
 */

// @lc code=start
function lengthOfLongestSubstring(s: string): number {
  let substrStart = 0;
  let maxLength = 0;
  let charToPos: Record<string, number> = {};

  for (let i = 0; i < s.length; i++) {
    const char = s.charAt(i);

    const prevPosition = charToPos[char];
    if (prevPosition !== undefined && prevPosition >= substrStart) {
      // 現在のsubstr内で重複が見つかれば、最初の文字から現在までが最も長い
      const currentLen = i - substrStart;
      if (currentLen > maxLength) maxLength = currentLen;
      substrStart = prevPosition + 1;
    }

    charToPos[char] = i;
  }

  if (s.length - substrStart > maxLength) {
    maxLength = s.length - substrStart;
  }

  return maxLength;
};
// @lc code=end

File: 3.longest-substring-without-repeating-characters.rb
#
# @lc app=leetcode id=3 lang=ruby
#
# [3] Longest Substring Without Repeating Characters
#

# @lc code=start
# @param {String} s
# @return {Integer}
def length_of_longest_substring(s)
  start = 0
  finish = 0

  longest = 0
  location = {} # char => index
  while finish < s.length
    if location.include?(s[finish]) && location[s[finish]] >= start
      current_length = finish - start
      longest = [longest, current_length].max

      start = location[s[finish]] + 1
    end

    location[s[finish]] = finish
    finish += 1
  end

  last_length = finish - start
  [longest, last_length].max
end
# @lc code=end

p(length_of_longest_substring('abcabcbb'))
p(length_of_longest_substring('pwwkew'))
p(length_of_longest_substring('abba'))
